### PR TITLE
Introduce BLOB_ID_V4 and routerBlobidCurrentVersion is still 3.

### PR DESCRIPTION
**Pull request by zzmao**
Introduce V4 only. No hack or change to V3. 

Looks like anyway we need this.

### CODE REVIEW

File: ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java [hunk 1]
```diff
@@ -396,7 +418,7 @@ public int compareTo(StoreKey o) {
     }
     BlobId other = (BlobId) o;
     int result = 0;
-    if (version < BLOB_ID_V3 || other.version < BLOB_ID_V3) {
```
Comment (vgkholla) [inline ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: this doesn't change

Comment (zzmao) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: Blob_ID_V3 and above would compare uuid only?

Comment (pnarayanan) [reply ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java]: yes, the javadoc for this method makes that clear.

File: ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java [hunk 2]
```diff
@@ -160,14 +160,19 @@ public void testComparisons() throws Exception {
...
         BlobId blobIdV3 = getRandomBlobId(BLOB_ID_V3);
+        BlobId blobIdV4 = getRandomBlobId(BLOB_ID_V4);
 
         assertTrue("blobIdV1 should be less than blobIdV2", blobIdV1.compareTo(blobIdV2) < 0);
         assertFalse(blobIdV1.equals(blobIdV2));
         assertTrue("blobIdV1 should be less than blobIdV3", blobIdV1.compareTo(blobIdV3) < 0);
         assertFalse(blobIdV1.equals(blobIdV3));
         assertTrue("blobIdV2 should be less than blobIdV3", blobIdV2.compareTo(blobIdV3) < 0);
         assertFalse(blobIdV2.equals(blobIdV3));
+        assertTrue("blobIdV3 should be less than blobIdV4", blobIdV3.compareTo(blobIdV4) < 0);
```
Comment (pnarayanan) [inline ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java]: for V3 and above, the comparison is to be done via the uuid only, so V3 < V4 is not always true.

File: ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java [hunk 3]
```diff
@@ -160,14 +160,19 @@ public void testComparisons() throws Exception {
...
+        assertFalse(blobIdV3.equals(blobIdV4));
 
+        assertTrue("blobIdV4 should be greater than blobIdV3", blobIdV4.compareTo(blobIdV3) > 0);
```
Comment (pnarayanan) [inline ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java]: same here, this should be omitted.

File: ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java [hunk 4]
```diff
@@ -245,7 +256,7 @@ public void testCrafting() throws Exception {
     assertEquals("Containers should match", crafted.getContainerId(), craftedAgain.getContainerId());
     assertEquals("The id string should match", crafted.getID(), craftedAgain.getID());
 
-    if (version == BLOB_ID_V3) {
+    if (version >= BLOB_ID_V3) {
```
Comment (pnarayanan) [inline ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java]: this test is not version specific, so the if condition was meant to be true only once.

File: ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java [hunk 5]
```diff
@@ -181,10 +182,17 @@ public void testComparisons() throws Exception {
         assertTrue(blobIdV2.equals(blobIdV2));
         assertTrue("blobIdV3 should be equal to blobIdV3", blobIdV3.compareTo(blobIdV3) == 0);
         assertTrue(blobIdV3.equals(blobIdV3));
+        assertTrue("blobIdV4 should be equal to blobIdV4", blobIdV4.compareTo(blobIdV4) == 0);
+        assertTrue(blobIdV4.equals(blobIdV4));
+
+        assertTrue("Comparison for blobIdV3 and above are based on uuid only",
+            blobIdV3.compareTo(blobIdV4) == blobIdV3.getUuid().compareTo(blobIdV4.getUuid()));
+        assertFalse(blobIdV3.equals(blobIdV4));
```
Comment (pnarayanan) [inline ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java]:
```java
assertTrue(blobIdV3.compareTo(blobIdV4) != blobIdV4.compareTo(blobIdV3));
```
tests non-equality, and also cross comparisons.

**Pull request merged**